### PR TITLE
fix for #125

### PR DESCRIPTION
**Pull request by tokejepsen**
- Fixing major bug of having switches new and old value around

### CODE REVIEW

**Comment by tokejepsen** [comment]
Using a callback to set the ```publish``` data member in the host. Don't know how clean that is, or whether its in the right place.

**Comment by mottosso** [comment]
Could you have a look at consolidating where this event is emitted, rather than doing it in multiple places? Something can easily be missed as it is now I think. Maybe the section toggle could make use of the instance and plug-in toggle methods somethow?

File: pyblish_qml/control.py [hunk 1]
```diff
@@ -17,6 +17,12 @@
 from pyblish_qml import settings
 
 
+def toggle_instance(instance, new_value, old_value):
+    instance.data['publish'] = bool(new_value)
+
+pyblish.api.register_callback("instanceToggled", toggle_instance)
```
Comment (mottosso) [inline pyblish_qml/control.py]: I'm not a fan of having modules produce side-effects on import. How about putting this in the [main](https://github.com/tokejepsen/pyblish-qml/blob/1be49764f34605d532b68d600f097e6275475a08/pyblish_qml/app.py#L226) instead, so we can safely import it without worry?

**Comment by tokejepsen** [comment]
Have a look at this now.

I'm slightly unsure of the "instanceAdded" callback solution, but its working.

**Comment by mottosso** [comment]
Having a look..

**Comment by mottosso** [comment]
Ah, the instanceAdded signal is currently coming from an internal class of the GUI. It doesn't have anything to do with Pyblish or publishing in general, but is just when the items are added visually to the GUI.

Did you have a usecase for this signal? It would find a better home in the Context class I think, possibly in thet `create_instance` method. But we shouldn't add signals unless there's a clear use for it. What do you think?

Otherwise, great work on the consolidation! It looks super smooth. I can't test it now, but by just eyeballing it I'd say it looks very maintainable and minimal.

**Comment by tokejepsen** [comment]
> It would find a better home in the Context class I think

Not quite getting where this Context class is? In pyblish-base?

**Comment by mottosso** [comment]
Yeah, sorry, I should have linked.

- [Context.create_instance](https://github.com/pyblish/pyblish-base/blob/a70c91c4620b5b4cb51cf749dc0de31f8fed909e/pyblish/plugin.py#L653)

**Comment by tokejepsen** [comment]
> Yeah, sorry, I should have linked.

But the point was to stay away from the pyblish-base repo, right? Only add the ```publish``` data member from pyblish-qml (some how), because it is needed here.

**Comment by mottosso** [comment]
> Only add the publish data member from pyblish-qml (some how), because it is needed here.

But this event isn't adding the `publish` data member? I thought it was just emitting whenever an instance got added to the Context?

Maybe let's start from the beginning. :) What is this event meant to do, and what is it for?

**Comment by tokejepsen** [comment]
> But this event isn't adding the publish data member? I thought it was just emitting whenever an instance got added to the Context?

Ohh, right yea. If we are just emitting the signal, we can receive it when using pyblish-qml, meaning nothing gets added unless the signal is received. I get it:)

**Comment by mottosso** [comment]
..now I don't get it. :)

Tell me, why this signal? It wasn't there before, why bother adding it?

**Comment by tokejepsen** [comment]
> Tell me, why this signal? It wasn't there before, why bother adding it?

To add the ```publish``` data member to all the instances; https://github.com/pyblish/pyblish-base/pull/244#issuecomment-179720959

**Comment by mottosso** [comment]
Ahaa, now I get you, you still want that to be there. I thought we agreed to not put data members in instances unnecessarily.

I think this is fine.

```python
if instance.data.get("publish", True):
    print "Instance is publishable"
```

What's wrong with that?

**Comment by tokejepsen** [comment]
> What's wrong with that?

I think its confusing for new users, that you set the ```publish``` data member to ```False``` when untoggling by default, and then when you query for it, the data member doesn't exist.
I don't mind using this workflow, but I just recall my first intuition to query the toggle state of an instance in the GUI, and it was to look at the ```publish``` data member, which was why I issued #125.

**Comment by mottosso** [comment]
Ah, ok. Sorry, I'm a bit slow today. :)

I can see what you mean, but let me know what you think about this, as this is the source of inspiration for I'm imagining things.

In Sublime Text, you have a default json configuration file for each feature, and an optional user-defined json file with overrides. The user-defined json file can contain every key in the default file, but only ever needs to contains what differs. In general, it only has one or two keys, whereas the default one has tens if not hundreds of potential targets.

I kind of like that workflow. It means they can change their default around, and your changes are just "layered" on top of those.

I've seen at least a few programs do things similarly, and it's how the configuration of Pyblish (when we were using json files to configure paths and such) was also working. But it does have the disadvantage of having users add entirely new keys to their user config, which can understandably be daunting.

I can see it both ways, what are your thoughts?

**Comment by tokejepsen** [comment]
> I can see it both ways, what are your thoughts?

I'm probably not entirely following you trail of thought. This "layered" approach sounds good, but the problem is we don't have a default data member for the toggle state of an instance in the GUI, so there is nothing to override.
This is actually what confused me to begin with, that you add a data member for the GUI to recognize the default toggle state of an instance. I know now that we are making an assumption, but my first thoughts were; "How does the GUI know which toggle state an instance is in, if there is no data member to begin with?"

**Comment by mottosso** [comment]
Let me have a think on this. The default I'm imagining would come from documentation, but there isn't any yet. Something like [this](https://github.com/pyblish/pyblish.api/wiki/Plugin). Listing data the GUI is looking for, along with what it considers default.

**Pull request merged**